File: shared/retry.py
"""Shared retry utility with exponential backoff.

Wraps any async function with automatic retry on transient Azure/HTTP failures.
"""

import asyncio
import logging
from typing import Any, Callable, Optional, Set

logger = logging.getLogger(__name__)

DEFAULT_RETRY_CODES: Set[int] = {429, 503, 504}


def is_retryable(exc: Exception, retry_codes: Set[int]) -> bool:
    """Check if an exception is retryable based on status code or message."""
    status = getattr(exc, "status_code", None) or getattr(exc, "code", None)
    if isinstance(status, int) and status in retry_codes:
        return True

    message = str(exc).lower()
    retryable_phrases = [
        "too many requests",
        "service unavailable",
        "gateway timeout",
        "connection reset",
        "timed out",
    ]
    return any(phrase in message for phrase in retryable_phrases)


async def with_retry(
    fn: Callable[..., Any],
    *args: Any,
    max_attempts: int = 3,
    base_delay_s: float = 1.0,
    max_delay_s: float = 30.0,
    retry_codes: Optional[Set[int]] = None,
    label: str = "operation",
    **kwargs: Any,
) -> Any:
    """Execute an async function with exponential backoff retry.

    Args:
        fn: Async callable to execute.
        *args: Positional arguments passed to fn.
        max_attempts: Maximum number of attempts (default 3).
        base_delay_s: Base delay in seconds (doubled each retry).
        max_delay_s: Maximum delay cap in seconds.
        retry_codes: HTTP status codes to retry on.
        label: Label for log messages.
        **kwargs: Keyword arguments passed to fn.

    Returns:
        Result of fn(*args, **kwargs).
    """
    if retry_codes is None:
        retry_codes = DEFAULT_RETRY_CODES

    last_error: Optional[Exception] = None

    for attempt in range(1, max_attempts + 1):
        try:
            return await fn(*args, **kwargs)
        except Exception as exc:
            last_error = exc

            if attempt >= max_attempts or not is_retryable(exc, retry_codes):
                raise

            delay = min(base_delay_s * (2 ** (attempt - 1)), max_delay_s)
            logger.warning(
                "Retry %d/%d for %s (delay=%.1fs): %s",
                attempt,
                max_attempts,
                label,
                delay,
                str(exc),
            )
            await asyncio.sleep(delay)

    raise last_error  # type: ignore[misc]

File: shared/retry.js
"use strict";

/**
 * retry.js — Shared retry utility with exponential backoff.
 *
 * Wraps any async function with automatic retry on transient failures.
 * Retries on HTTP 429, 503, 504 by default.
 */

/** @type {number[]} Default HTTP status codes to retry on */
const DEFAULT_RETRY_CODES = [429, 503, 504];

/**
 * Determine if an error is retryable based on HTTP status code.
 * @param {Error} err
 * @param {number[]} retryCodes
 * @returns {boolean}
 */
function isRetryable(err, retryCodes) {
  const status =
    err.statusCode || err.status || err.code || err.response?.status;
  if (typeof status === "number" && retryCodes.includes(status)) return true;

  // Azure SDK errors often have a `code` string
  const message = (err.message || "").toLowerCase();
  if (message.includes("too many requests")) return true;
  if (message.includes("service unavailable")) return true;
  if (message.includes("gateway timeout")) return true;
  if (message.includes("econnreset") || message.includes("etimedout")) return true;

  return false;
}

/**
 * Execute an async function with exponential backoff retry.
 *
 * @param {() => Promise<*>} fn - Async function to execute
 * @param {object} [options]
 * @param {number} [options.maxAttempts=3] - Maximum number of attempts
 * @param {number} [options.baseDelayMs=1000] - Base delay in ms (doubled each retry)
 * @param {number} [options.maxDelayMs=30000] - Maximum delay cap in ms
 * @param {number[]} [options.retryOn=[429,503,504]] - HTTP status codes to retry
 * @param {string} [options.label="operation"] - Label for log messages
 * @returns {Promise<*>} Result of fn()
 */
async function withRetry(fn, options = {}) {
  const {
    maxAttempts = 3,
    baseDelayMs = 1000,
    maxDelayMs = 30000,
    retryOn = DEFAULT_RETRY_CODES,
    label = "operation",
  } = options;

  let lastError;

  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (err) {
      lastError = err;

      if (attempt >= maxAttempts || !isRetryable(err, retryOn)) {
        throw err;
      }

      // Exponential backoff with jitter
      const delay = Math.min(
        baseDelayMs * Math.pow(2, attempt - 1) + Math.random() * 100,
        maxDelayMs
      );

      console.log(
        JSON.stringify({
          message: `Retry ${attempt}/${maxAttempts} for ${label}`,
          delay: Math.round(delay),
          error: err.message,
        })
      );

      await new Promise((resolve) => setTimeout(resolve, delay));
    }
  }

  throw lastError;
}

module.exports = { withRetry, isRetryable, DEFAULT_RETRY_CODES };
